cogs/stats/leaderboard.py

```diff
@@ -0,0 +1,127 @@
+        self.author = author
+        self.current_embed: LeaderboardEmbed = embed
+        super().__init__(timeout=300)
+
+    async def interaction_check(self, interaction: discord.Interaction):  # type: ignore
+        if interaction.user != self.author:
+            return await interaction.response.send_message("This is not your view!", ephemeral=True)
+
+        return True
+
+    @discord.ui.button(style=discord.ButtonStyle.secondary, label="All Time", disabled=True)
+    async def all_time_callback(self, interaction: discord.Interaction[HideoutManager], button: discord.ui.Button):  # type: ignore
+        for btn in self.children:
+            if isinstance(btn, discord.ui.Button):
+                btn.disabled = False
+
+        button.disabled = True
+
+        embed = await self.current_embed.update_leaderboard(interval=None)
+
+        await interaction.edit_original_response(embed=embed, view=self)
+
+    @discord.ui.button(style=discord.ButtonStyle.secondary, label="Last 30 Days")
+    async def _30_day_callback(self, interaction: discord.Interaction[HideoutManager], button: discord.ui.Button):  # type: ignore
+        for btn in self.children:
+            if isinstance(btn, discord.ui.Button):
+                btn.disabled = False
+
+        button.disabled = True
+
+        embed = await self.current_embed.update_leaderboard(interval="30 DAYS")
+
+        await interaction.edit_original_response(embed=embed, view=self)
+
+    @discord.ui.button(style=discord.ButtonStyle.secondary, label="Last 7 Days")
+    async def _7_day_callback(self, interaction: discord.Interaction[HideoutManager], button: discord.ui.Button):  # type: ignore
+        for btn in self.children:
+            if isinstance(btn, discord.ui.Button):
+                btn.disabled = False
+
+        button.disabled = True
+
+        embed = await self.current_embed.update_leaderboard(interval="7 DAYS")
+
+        await interaction.response.edit_message(embed=embed, view=self)
+
+
+class LeaderboardEmbed(discord.Embed):
+    def __init__(self, pool: asyncpg.Pool[asyncpg.Record], bot: HideoutManager):
+        self._pool = pool
+        self._bot = bot
+        super().__init__(title="Leaderboard", color=discord.Color.from_str("#1b1d21"))
+
+    async def update_leaderboard(self, interval: str | None) -> discord.Embed:
+        query = """
+        SELECT author_id, COUNT(*) as message_count FROM message_info 
+        WHERE deleted = FALSE 
+        AND is_bot = $1
+        {0}
+        GROUP BY author_id 
+        ORDER BY message_count DESC LIMIT 10
+        """
+        self._data: list[asyncpg.Record] = await self._pool.fetch(
+            query.format("--" if interval is None else f"AND created_at > NOW - INTERVAL {interval}"), False
+        )
+
+        if not self._data:
+            raise RuntimeError("No leaderboard can be generated.")
+
+        for rank, user in enumerate(self._data):
```

Comment:
Shouldn't be 0-indexed
```suggestion
        for rank, user in enumerate(self._data, start=1):
```